### PR TITLE
Completed drivetrain

### PR DESCRIPTION
**Pull request by a1cd**

### CODE REVIEW

File: src/main/kotlin/frc/robot/Constants.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package frc.robot
+
+import kotlin.math.PI
+
+object Constants {
+    const val gyroReversed = false
```
Comment (Ryan-Bauroth) [inline src/main/kotlin/frc/robot/Constants.kt:6]: const val gyroReversed seems like a test value/not important to the build

File: src/main/kotlin/frc/robot/Robot.kt [hunk 2]
```diff
@@ -1,111 +1,29 @@
...
 import edu.wpi.first.wpilibj.TimedRobot
-import edu.wpi.first.wpilibj.smartdashboard.SendableChooser
-import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
+import edu.wpi.first.wpilibj2.command.CommandScheduler
 
 /**
  * The VM is configured to automatically run this class, and to call the
  * functions corresponding to each mode, as described in the TimedRobot
- * documentation. If you change the name of this class or the package after
+ * documentation. If you change the mname of this class or the package after
```
Comment (25lid) [inline src/main/kotlin/frc/robot/Robot.kt:9]: spelling mistake

File: src/main/kotlin/frc/robot/commands/DriveMotorTest.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        println(controlScheme)
+        println(controlScheme.driveTest.asBoolean)
+        println(controlScheme.turnTest.asBoolean)
+        println(controlScheme.testPercent)
+
+    }
+
+    override fun execute() {
+//        modules.forEach {
+//            it.driveMotor.set(controlScheme.testPercent)
```
Comment (Ryan-Bauroth) [inline src/main/kotlin/frc/robot/commands/DriveMotorTest.kt:32]: this also looks like test code?

File: src/main/kotlin/frc/robot/commands/TurnMotorTest.kt [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    override fun end(interrupted: Boolean) {
+        modules.forEach {
+            it.driveMotor.set(0.0)
+        }
+    }
+
+    override fun initSendable(builder: SendableBuilder?) {
+        super.initSendable(builder)
```
Comment (Ryan-Bauroth) [inline src/main/kotlin/frc/robot/commands/TurnMotorTest.kt:41]: //TODO delete

**Review by 25lid — APPROVED** [review]

File: src/main/kotlin/frc/robot/Constants.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package frc.robot
+
+import kotlin.math.PI
+
+object Constants {
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/Constants.kt:5]: Might want to add comments to each of these constants to describe what they do

File: src/main/kotlin/frc/robot/subsystems/Drivetrain.kt [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        position = Translation2d(Constants.MODULE_DISTANCE_X / 2, Constants.MODULE_DISTANCE_Y / 2),
+        controlScheme = controlScheme
+    ) // FIXME: change postion to new drivebase measurements
+    val frontRight = SwerveModule( // backleft
+        Constants.FRDriveMotorId,
+        FRTurnMotorId,
+        FRTurnEncoderId,
+        "frontRight",
+        Translation2d(Constants.MODULE_DISTANCE_X / 2, -Constants.MODULE_DISTANCE_Y / 2),
+        angleZero = 73.0 + 3.1,
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/subsystems/Drivetrain.kt:56]: Might want to move to constants

File: src/main/kotlin/frc/robot/subsystems/Drivetrain.kt [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                desiredStates[i]?.speedMetersPerSecond ?: 0.0,
+                desiredStates[i]?.angle ?: Rotation2d()
+            )
+        }
+    }
+
+    /**
+     *  Resets the drive encoders to currently read a position of 0
+     */
+    fun resetEncoders() {
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/subsystems/Drivetrain.kt:169]: This doesn't reset front right

File: src/main/kotlin/frc/robot/subsystems/Drivetrain.kt [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    )
+
+//    private val gyro: Gyro = ADXRS450_Gyro()
+
+    // Odometry class for tracking robot pose
+    var odometry = SwerveDriveOdometry(
+        kinematics,
+        Rotation2d(),//gyro.rotation2d,
+        modules
+            .map { it.position.toSwerveModulePosition() }
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/subsystems/Drivetrain.kt:92]: SwerveModulePosition isn't the position of the module on the robot, but the distance each module's driving motor has traveled. Should be SwerveModulePosition(driveMotor's encoder position in meters, module angle rotation2d)

> The initial positions of the swerve modules (as an array of SwerveModulePosition). In Java, this must be constructed with each wheel position in meters. 

The physical positions (on the robot) of each swerve module are already passed as part of the kinematics object.

File: PhotonLib-json-1.0.json [hunk 9]
```diff
@@ -0,0 +1,41 @@
+{
```
Comment (trevnels) [inline PhotonLib-json-1.0.json:1]: This file should go in the vendordeps folder

File: src/main/kotlin/frc/robot/subsystems/SwerveModule.kt [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+    // motor sets
+    val driveMotorSetEntry = tab.add("$mname Drive Motor Set", 0.0).entry
+    val turnMotorSetEntry = tab.add("$mname Turn Motor Set", 0.0).entry
+
+    // position/angle
+    val positionEntry = tab.add("$mname Position", 0.0).entry
+    val angleEntry = tab.add("$mname Angle", 0.0).entry
+
+    val DRIVE_P = 2.37
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/subsystems/SwerveModule.kt:76]: PID/FF constants probably belong in constants

File: src/main/kotlin/frc/robot/subsystems/SwerveModule.kt [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    val drivePid = ProfiledPIDController(
+        DRIVE_P,
+        DRIVE_I,
+        DRIVE_D,
+        TrapezoidProfile.Constraints(2.0, .0)// TODO: Fix these
+    )
+    private val driveFF = SimpleMotorFeedforward(0.21862, 2.2997, 0.26242)
+
+
+    val ANGLE_P = 0.5
```
Comment (trevnels) [inline src/main/kotlin/frc/robot/subsystems/SwerveModule.kt:88]: This P value might be too low. I think we should double-check the sysid value and make sure it's in position mode

**Review by trevnels — COMMENTED** [review]
Few comments, overall looks good

**Review by trevnels — APPROVED** [review]

**Review by Ryan-Bauroth — APPROVED** [review]

**Pull request merged**